### PR TITLE
feat: xml-json-conversion plugin convert xml data from request body to json response, and vice versa

### PR DESCRIPTION
**Pull request by mangoGoForward**
### Pre-submission checklist:

* [ ] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**
## Summary

- [**Name**](#name)
- [**Attributes**](#Attributes)
- [**How To Enable**](#how-to-enable)
- [**Test Plugin**](#test-plugin)
- [**Disable Plugin**](#disable-plugin)

## name

xml-json-conversion plugin convert xml data from request body to json response, and vice versa

## Attributes

None

## how-to-enable

Here's an example, enable this plugin on the specified route:

```shell
curl -i http://127.0.0.1:9080/apisix/admin/routes/1  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "/hello",
    "plugins": {
        "xml-json-conversion": {}
    },
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "39.97.63.215:80": 1
        }
    }
}'
```

## test-plugin

Via curl to access

```shell
curl -X GET http://127.0.0.1:9080/hello \
-H 'Content-Type: text/xml' \
-H 'Accept: application/json' \
--data '
<people>
  <person>
    <name>Manoel</name>
    <city>Palmas-TO</city>
  </person>
</people>'


{"people":{"person":{"name":"Manoel","city":"Palmas-TO"}}}
```

## disable-plugin

When you want to disable the `xml-json-conversion` plugin, it is very simple,
you can delete the corresponding json configuration in the plugin configuration,
no need to restart the service, it will take effect immediately:

```shell
curl -i http://127.0.0.1:9080/apisix/admin/routes/1  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "/hello",
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "39.97.63.215:80": 1
        }
    }
}'
```

### CODE REVIEW

**Comment by tokers** [comment]
@mangoGoForward The PR title is not semantic, please change it so that the semantic pull request action can pass.

File: apisix/plugins/xml-json-conversion.lua [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    parser:parse(xml_data)
+    return 200, cjson.encode(convertHandler.root)
+end
+
+local function json2xml(table_data)
+    return 200, xml2lua.toXml(cjson.decode(table_data))
+end
+
+function _M.access(conf, ctx)
+    core.log.warn("\n||||||| ctx ,", core.json.encode(ctx, true))
```
Comment (membphis) [inline apisix/plugins/xml-json-conversion.lua:72]: why do we need `\n|||||||`?

File: apisix/plugins/xml-json-conversion.lua [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    core.log.warn("\n||||||| conf ,", core.json.encode(conf, true))
+
+    local request_header = ngx.req.get_headers()
+    ngx.req.read_body()
+    local request_body = ngx.req.get_body_data()
```
Comment (membphis) [inline apisix/plugins/xml-json-conversion.lua:77]: call the `apisix.core` is easier

File: apisix/plugins/xml-json-conversion.lua [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    if nil == request_body then
+        local file_name = ngx.req.get_body_file()
+        if file_name then
+            request_body = getFile(file_name)
+        end
+    end
+
+    core.log.warn("------- request_body : ", request_body)
```
Comment (membphis) [inline apisix/plugins/xml-json-conversion.lua:85]: should remove this line

File: apisix/plugins/xml-json-conversion.lua (hunk 1 above)
Comment (mangoGoForward) [reply apisix/plugins/xml-json-conversion.lua:72]: Sorry to reply to you so late, I'll delete those lines, Thanks for your guidance

File: apisix/plugins/xml-json-conversion.lua (hunk 2 above)
Comment (mangoGoForward) [reply apisix/plugins/xml-json-conversion.lua]: Great idea

File: apisix/plugins/xml-json-conversion.lua [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+local core    = require("apisix.core")
+local xml2lua = require("xml2lua")
+local handler = require("xmlhandler.tree")
+local cjson   = require('cjson.safe')
+local string  = string
+
+local metadata_schema = {
+    type = "object",
+    properties = {},
+    additionalProperties = false,
```
Comment (tokers) [inline apisix/plugins/xml-json-conversion.lua:27]: Now we don't set constraints on it.

**Comment by tokers** [comment]
It seems that we can extend this plug-in, so hat it can be extensible if new formats need to be supported.

We may let users to specify the input and output type.

File: apisix/plugins/xml-json-conversion.lua (hunk 4 above)
Comment (mangoGoForward) [reply apisix/plugins/xml-json-conversion.lua:27]: It seems unnecessary, I will remove

**Comment by mangoGoForward** [comment]
> It seems that we can extend this plug-in, so hat it can be extensible if new formats need to be supported.
> 
> We may let users to specify the input and output type.
Thanks

**Comment by mangoGoForward** [comment]
> It seems that we can extend this plug-in, so hat it can be extensible if new formats need to be supported.
> 
> We may let users to specify the input and output type.


@tokers Thanks

File: apisix/plugins/xml-json-conversion.lua (hunk 1 above)
Comment (mangoGoForward) [reply apisix/plugins/xml-json-conversion.lua]: done

File: apisix/plugins/xml-json-conversion.lua (hunk 2 above)
Comment (mangoGoForward) [reply apisix/plugins/xml-json-conversion.lua]: done

File: apisix/plugins/xml-json-conversion.lua (hunk 3 above)
Comment (mangoGoForward) [reply apisix/plugins/xml-json-conversion.lua]: done

**Comment by mangoGoForward** [comment]
> @mangoGoForward The PR title is not semantic, please change it so that the semantic pull request action can pass.

Thanks, Already change

File: apisix/plugins/xml-json-conversion.lua (hunk 4 above)
Comment (mangoGoForward) [reply apisix/plugins/xml-json-conversion.lua]: done

**Comment by mangoGoForward** [comment]
> We may let users to specify the input and output type.

Do you have any good suggestions? And We need users to specify request "Content-Type" and "Accept" in this plugin
https://github.com/mangoGoForward/apisix/blob/df782db7905cfe983188e53e2f645db48d956eed/apisix/plugins/xml-json-conversion.lua#L91

**Comment by tokers** [comment]
> > We may let users to specify the input and output type.
> 
> Do you have any good suggestions? And We need users to specify "Content-Type" and "Accept" of request headers in this plugin
> https://github.com/mangoGoForward/apisix/blob/df782db7905cfe983188e53e2f645db48d956eed/apisix/plugins/xml-json-conversion.lua#L91

Such as:

```json
{
  "rules": [
     {
        "from": "json",
        "to": "yaml"
     },
     {
        "from": "yaml",
         "to": "json"
     }
  ]
}
```

**Comment by mangoGoForward** [comment]
> > > We may let users to specify the input and output type.
> > 
> > 
> > Do you have any good suggestions? And We need users to specify "Content-Type" and "Accept" of request headers in this plugin
> > https://github.com/mangoGoForward/apisix/blob/df782db7905cfe983188e53e2f645db48d956eed/apisix/plugins/xml-json-conversion.lua#L91
> 
> Such as:
> 
> ```json
> {
>   "rules": [
>      {
>         "from": "json",
>         "to": "yaml"
>      },
>      {
>         "from": "yaml",
>          "to": "json"
>      }
>   ]
> }
> ```

Do you mean support metadata like this:

```
local schema = {
    type = "object",
    properties = {
        from = {
            type = "string",
            default = "yaml"
        },
        to = {
            type = "string",
            default = "json"
        }
    },
    required = {"from", "to"}
}
```
When user create Apisix Route or Service, they can  specify the input and output type?

**Comment by tokers** [comment]
> > > > We may let users to specify the input and output type.
> > > 
> > > 
> > > Do you have any good suggestions? And We need users to specify "Content-Type" and "Accept" of request headers in this plugin
> > > https://github.com/mangoGoForward/apisix/blob/df782db7905cfe983188e53e2f645db48d956eed/apisix/plugins/xml-json-conversion.lua#L91
> > 
> > 
> > Such as:
> > ```json
> > {
> >   "rules": [
> >      {
> >         "from": "json",
> >         "to": "yaml"
> >      },
> >      {
> >         "from": "yaml",
> >          "to": "json"
> >      }
> >   ]
> > }
> > ```
> 
> Do you mean support metadata like this:
> 
> ```
> local schema = {
>     type = "object",
>     properties = {
>         from = {
>             type = "string",
>             default = "yaml"
>         },
>         to = {
>             type = "string",
>             default = "json"
>         }
>     },
>     required = {"from", "to"}
> }
> ```
> 
> When user create Apisix Route or Service, they can specify the input and output type?

Yep, then this plugin will be more flexible.

File: apisix/plugins/xml-json-conversion.lua [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+function _M.access(conf, ctx)
+    local request_header = core.request.headers()
+    local req_body, err = core.request.get_body()
+    if err or req_body == nil or req_body == '' then
+        core.log.error("failed to read request body: ", err)
+        core.response.exit(400, {error_msg = "invalid request body: " .. err})
+    end
+
+    if request_header["Content-Type"] == "application/json" then
```
[the remark this answers is not in the document]
Comment (mangoGoForward) [reply apisix/plugins/xml-json-conversion.lua:62]: I want to support metadata to let users can specify input and output type, and should I  also need to focus on user's request header?

[the remark this answers is not in the document]
Comment (tzssangglass) [reply apisix/plugins/xml-json-conversion.lua:62]: I think it's unnecessary, just to care if it's the `Content-Type` defined by the metadata that needs to be processed

[the remark this answers is not in the document]
Comment (mangoGoForward) [reply apisix/plugins/xml-json-conversion.lua:62]: Thanks for your opinion, I will change this plugin later

File: apisix/plugins/xml-json-conversion.lua [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+end
+
+function _M.access(conf, ctx)
+    local request_header = core.request.headers()
+    local req_body, err = core.request.get_body()
+    if err or req_body == nil or req_body == '' then
+        core.log.error("failed to read request body: ", err)
+        core.response.exit(400, {error_msg = "invalid request body: " .. err})
+    end
+
```
[the remark this answers is not in the document]
Comment (mangoGoForward) [reply apisix/plugins/xml-json-conversion.lua]: done

File: apisix/plugins/xml-json-conversion.lua (hunk 5 above)
[the remark this answers is not in the document]
Comment (mangoGoForward) [reply apisix/plugins/xml-json-conversion.lua]: Already changed, could you please give me some suggestion ? @tokers @tzssangglass

File: apisix/plugins/xml-json-conversion.lua [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+--function _M.control_api()
+--    return {
+--        {
+--            methods = {"GET"},
+--            uris ={"/v1/plugin/xml-json-conversion"},
+--            handler = get_json,
+--        }
+--    }
+--end
```
Comment (tzssangglass) [inline apisix/plugins/xml-json-conversion.lua:162]: just curious, what is this comments for?

File: t/plugin/xml-json-conversion.t [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+            ngx.say(body)
+        }
+    }
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
```
Comment (tzssangglass) [inline t/plugin/xml-json-conversion.t:157]: Can you add validation of the content after the xml to json conversion?

**Review by tzssangglass — COMMENTED** [review]
json to xml test cases are also missing

**Comment by tzssangglass** [comment]
others LGTM

**Comment by mangoGoForward** [comment]
> json to xml test cases are also missing

Use request query string of "from" and "to" in function of get_json, like "from" is "xml", and "to" is "json", it can

**Comment by mangoGoForward** [comment]
> json to xml test cases are also missing

Use request query string

File: apisix/plugins/xml-json-conversion.lua (hunk 7 above)
Comment (mangoGoForward) [reply apisix/plugins/xml-json-conversion.lua:162]: Just registered a control api for test, and do not create a route or service

Comment (mangoGoForward) [reply apisix/plugins/xml-json-conversion.lua]: done, Already deleted

File: t/plugin/xml-json-conversion.t (hunk 8 above)
Comment (mangoGoForward) [reply t/plugin/xml-json-conversion.t:157]: done, and also add json to xml test cases

**Comment by mangoGoForward** [comment]
Badly, I merged upstream master changes, and deleted fork develop branch for revert there changes, then cause this PR forced closed. can I recreate a PR for commit?

**Comment by tzssangglass** [comment]
> Badly, I merged upstream master changes, and deleted fork develop branch for revert there changes, then cause this PR forced closed. can I recreate a PR for commit?

yes

**Pull request closed without merging**